go/ct/driver/probe.go

```diff
@@ -0,0 +1,198 @@
+//
+// Copyright (c) 2024 Fantom Foundation
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at fantom.foundation/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use
+// of this software will be governed by the GNU Lesser General Public Licence v3
+//
+
+package main
+
+import (
+	"fmt"
+	"runtime"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"pgregory.net/rand"
+
+	"github.com/Fantom-foundation/Tosca/go/ct"
+	"github.com/Fantom-foundation/Tosca/go/ct/common"
+	cliUtils "github.com/Fantom-foundation/Tosca/go/ct/driver/cli"
+	"github.com/Fantom-foundation/Tosca/go/ct/gen"
+	"github.com/Fantom-foundation/Tosca/go/ct/rlz"
+	"github.com/Fantom-foundation/Tosca/go/ct/spc"
+	"github.com/Fantom-foundation/Tosca/go/ct/st"
+	"github.com/dsnet/golib/unitconv"
+	"github.com/urfave/cli/v2"
+	"golang.org/x/exp/maps"
+)
+
+var ProbeCmd = cli.Command{
+	Action:    doProbe,
+	Name:      "probe",
+	Usage:     "Run random tests on an EVM implementation",
+	ArgsUsage: "<EVM>",
+	Flags: []cli.Flag{
+		cliUtils.FilterFlag,
+		cliUtils.SeedFlag,
+		&cli.IntFlag{
+			Name:  "timeout",
+			Usage: "maximum time in minutes to run the probing. Default 30 minutes",
+			Value: 30,
+		},
+	},
+}
+
+func doProbe(context *cli.Context) error {
+	var evmIdentifier string
+	if context.Args().Len() >= 1 {
+		evmIdentifier = context.Args().Get(0)
+	}
+
+	evm, ok := evms[evmIdentifier]
+	if !ok {
+		return fmt.Errorf("invalid EVM identifier, use one of: %v", maps.Keys(evms))
+	}
+
+	jobCount := cliUtils.JobsFlag.Fetch(context)
+	if jobCount <= 0 {
+		jobCount = runtime.NumCPU()
+	}
+
+	seed := cliUtils.SeedFlag.Fetch(context)
+
+	timeout := time.Duration(context.Int("timeout")) * time.Minute
+	hasTimeouted := false
+
+	// The constraints to be placed on generated states.
+	condition := rlz.And(
+		rlz.IsCode(rlz.Pc()),
+		// update to cancun when all necessary changed have been implemented
+		rlz.RevisionBounds(common.MinRevision, common.R12_Shanghai),
+	)
+
+	fmt.Printf("Start random tests on %s using %d jobs, seed %d, timeout %v and constraints %s ...\n", evmIdentifier, jobCount, seed, timeout, condition)
+
+	// Run a progress printer in the background.
+	counter := atomic.Uint64{}
+	stopProgressPrinter := make(chan struct{})
+	var progressGroup sync.WaitGroup
+	progressGroup.Add(1)
+	go func() {
+		defer progressGroup.Done()
+		start := time.Now()
+		last := uint64(0)
+		for {
+			select {
+			case <-stopProgressPrinter:
+				return
+			case <-time.After(5 * time.Second):
```

Comment:
Consider using a [Ticker](https://gobyexample.com/tickers) for this. Tickers are made for this use case.